src/message/__tests__/fetchActions-test.js

```diff
@@ -102,69 +102,131 @@ describe('fetchActions', () => {
-      await expect(tryFetchFunc.mock.results[0].value).resolves.toBe('hello');
+          expect(tryFetchFunc).toHaveBeenCalledTimes(1);
+          await expect(tryFetchFunc.mock.results[0].value).resolves.toBe('hello');
 
-      jest.runAllTimers();
-    });
+          jest.runAllTimers();
+        });
 
-    // TODO: test more errors, like regular `new Error()`s. Unexpected
-    // errors should actually cause the retry loop to break; we'll fix
-    // that soon.
-    test('retries a call if there is a non-client error', async () => {
-      const serverError = new ApiError(500, {
-        code: 'SOME_ERROR_CODE',
-        msg: 'Internal Server Error',
-        result: 'error',
-      });
+        test('Rethrows a 4xx error without retrying', async () => {
+          const apiError = new ApiError(400, {
+            code: 'BAD_REQUEST',
+            msg: 'Bad Request',
+            result: 'error',
+          });
 
-      // fail on first call, succeed second time
-      let callCount = 0;
-      const thrower = jest.fn(() => {
-        callCount++;
-        if (callCount === 1) {
-          throw serverError;
-        }
-        return 'hello';
-      });
+          const func = jest.fn(async () => {
+            throw apiError;
+          });
 
-      const tryFetchFunc = jest.fn(async () => {
-        await fakeSleep(10);
-        return thrower();
-      });
+          await expect(tryFetch(func, withRetry)).rejects.toThrow(apiError);
+          expect(func).toHaveBeenCalledTimes(1);
 
-      await expect(tryFetch(tryFetchFunc)).resolves.toBe('hello');
+          jest.runAllTimers();
+        });
 
-      expect(tryFetchFunc).toHaveBeenCalledTimes(2);
-      await expect(tryFetchFunc.mock.results[0].value).rejects.toThrow(serverError);
-      await expect(tryFetchFunc.mock.results[1].value).resolves.toBe('hello');
+        test('Rethrows an unexpected error without retrying', async () => {
+          const unexpectedError = new Error('You have displaced the mirth.');
 
-      jest.runAllTimers();
-    });
+          const func = jest.fn(async () => {
+            throw unexpectedError;
+          });
 
-    test('Rethrows a 4xx error without retrying', async () => {
-      const apiError = new ApiError(400, {
-        code: 'BAD_REQUEST',
-        msg: 'Bad Request',
-        result: 'error',
+          await expect(tryFetch(func, withRetry)).rejects.toThrow(unexpectedError);
+          expect(func).toHaveBeenCalledTimes(1);
+
+          jest.runAllTimers();
+        });
+
+        test('times out after hanging on one request', async () => {
+          const tryFetchPromise = tryFetch(async () => {
+            await new Promise((resolve, reject) => {});
+          }, withRetry);
+
+          await fakeSleep(60000);
+          return expect(tryFetchPromise).rejects.toThrow(TimeoutError);
+        });
+      },
+    );
+
+    describe('if asked to retry', () => {
+      test('retries a call if there is a recoverable error', async () => {
```

Comment:
fetchActions: Fail early on all non–known-retryable errors in `tryFetch`.

I appreciate your careful use of both an en-dash and a hyphen. Wouldn't want a reader to think we're talking about errors that are retryable in a non-known fashion :-P

Reply:
I wish I still had access to the Chicago Manual of Style; it looks like this would be covered in Chapter 6: https://www.chicagomanualofstyle.org/16/ch06/ch06_toc.html 😛 (but I no longer have my college-sponsored subscription, alas).